indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisorStateManager.java

```diff
@@ -0,0 +1,346 @@
+    UNHEALTHY_TASKS(2),
+    UNABLE_TO_CONNECT_TO_STREAM(3),
+    LOST_CONTACT_WITH_STREAM(4),
+    // Non-error states are equal priority
+    WAITING_TO_RUN(5),
+    CONNECTING_TO_STREAM(5),
+    DISCOVERING_INITIAL_TASKS(5),
+    CREATING_TASKS(5),
+    RUNNING(5),
+    SUSPENDED(5),
+    SHUTTING_DOWN(5);
+
+    // Lower priority number means higher priority and vice versa
+    private final int priority;
+
+    State(int priority)
+    {
+      this.priority = priority;
+    }
+
+    // We only want to set these if the supervisor hasn't had a successful iteration yet
+    public boolean isOnlySetWhenNoSuccessfulRunYet()
+    {
+      Set<State> firstRunStates = ImmutableSet.of(
+          WAITING_TO_RUN,
+          CONNECTING_TO_STREAM,
+          DISCOVERING_INITIAL_TASKS,
+          CREATING_TASKS,
+          RUNNING
+      );
+      return firstRunStates.contains(this);
+    }
+
+    public boolean isHealthy()
+    {
+      Set<State> unhealthyStates = ImmutableSet.of(
+          UNHEALTHY_SUPERVISOR,
+          UNHEALTHY_TASKS,
+          UNABLE_TO_CONNECT_TO_STREAM,
+          LOST_CONTACT_WITH_STREAM
+      );
+      return !unhealthyStates.contains(this);
+    }
+  }
+
+  private State supervisorState;
+  private final int healthinessThreshold;
+  private final int unhealthinessThreshold;
+  private final int healthinessTaskThreshold;
+  private final int unhealthinessTaskThreshold;
+
+  private boolean atLeastOneSuccessfulRun = false;
+  private boolean currentRunSuccessful = true;
+  private final CircularBuffer<TaskState> completedTaskHistory;
+  private final CircularBuffer<State> stateHistory; // From previous runs
+  private final ExceptionEventStore eventStore;
+
+  public SeekableStreamSupervisorStateManager(
+      State initialState,
+      SeekableStreamSupervisorConfig supervisorConfig
+  )
+  {
+    this.supervisorState = initialState;
+    this.healthinessThreshold = supervisorConfig.getHealthinessThreshold();
+    this.unhealthinessThreshold = supervisorConfig.getUnhealthinessThreshold();
+    this.healthinessTaskThreshold = supervisorConfig.getTaskHealthinessThreshold();
+    this.unhealthinessTaskThreshold = supervisorConfig.getTaskUnhealthinessThreshold();
+    Preconditions.checkArgument(
+        supervisorConfig.getMaxStoredExceptionEvents() >=
+        Math.max(healthinessThreshold, unhealthinessThreshold),
+        "numExceptionEventsToStore must be greater than or equal to both "
+        + "healthinessThreshold and unhealthinessThreshold"
+    );
+    this.eventStore = new ExceptionEventStore(
+        supervisorConfig.getMaxStoredExceptionEvents(),
+        supervisorConfig.isStoringStackTraces()
+    );
+    this.completedTaskHistory = new CircularBuffer<>(Math.max(healthinessTaskThreshold, unhealthinessTaskThreshold));
+    this.stateHistory = new CircularBuffer<>(Math.max(healthinessThreshold, unhealthinessThreshold));
+  }
+
+  public Optional<State> setStateAndCheckIfFirstRun(State newState)
```

Comment:
I think it's more readable if this is just called `setState`, the logic around some states only occurring on the first iteration of the supervisor could be documented more fully in a method javadoc or somewhere else state-related